core/src/main/java/org/apache/calcite/jdbc/JavaTypeFactoryImpl.java

```diff
@@ -244,10 +244,11 @@ public RelDataType toSql(RelDataType type) {
   public static RelDataType toSql(final RelDataTypeFactory typeFactory,
       RelDataType type) {
     if (type instanceof RelRecordType) {
-      return typeFactory.createStructType(
-          Lists.transform(type.getFieldList(),
-              field -> toSql(typeFactory, field.getType())),
-          type.getFieldNames());
+      return typeFactory.createTypeWithNullability(
```

Comment:
How many times are we going to visit each field? I have the impression that this recursion will visit attributes multiple times for no reason but maybe I am wrong.

Reply:
Another possibility will be avoiding the `createTypeWithNullability` call and create the struct type directly with the appropriate nullability via the typeFactory:
```
typeFactory.createStructType(
    Lists.transform(type.getFieldList(), field -> toSql(typeFactory, field.getType())),
    type.getFieldNames(),
    type.isNullable());
```
This it totally possible, but it would require to add a new method to RelDataTypeFactory interface:
```
private RelDataType createStructType(StructKind kind,
      final List<RelDataType> typeList,
      final List<String> fieldNameList,
      final boolean nullable)
```
This method is already defined (though for the moment as `private`) in RelDataTypeFactoryImpl.

Reply:
Since for the moment there is no other createType method who takes nullability as a parameter I would say not to add a new one in the public API. On the other hand maybe we can avoid the problem of multiple traversals by adding a new 
`private static RelDataType toSql(final RelDataTypeFactory typeFactory, RelDataType type, boolean changeNullability) `
with which we can control when we have to change nullability. In any case only the top-level nullability is important for structs since the nullability of nested fields can be erased by the current code in createTypeWithNullability.

What do you think?

Reply:
Ok, modified with new private method to change nullability only at "root" level.